lib/utils/compose.coffee

```diff
@@ -0,0 +1,747 @@
+			binPath = qemu.qemuPathInContext(path.join(projectPath, task.context))
+			transpose.transposeTarStream task.buildStream,
+				hostQemuPath: toPosixPath(binPath)
+				containerQemuPath: "/tmp/#{qemu.QEMU_BIN_NAME}"
+			.then (stream) ->
+				task.buildStream = stream
+			.return([ task, binPath ])
+	.map ([ task, qemuPath ]) ->
+		Promise.resolve(task).tap (task) ->
+			captureStream = buildLogCapture(task.external, task.logBuffer)
+
+			if task.external
+				# External image -- there's no build to be performed,
+				# just follow pull progress.
+				captureStream.pipe(task.logStream)
+				task.progressHook = pullProgressAdapter(captureStream)
+			else
+				task.streamHook = (stream) ->
+					if qemuPath?
+						buildThroughStream = transpose.getBuildThroughStream
+							hostQemuPath: toPosixPath(qemuPath)
+							containerQemuPath: "/tmp/#{qemu.QEMU_BIN_NAME}"
+						rawStream = stream.pipe(buildThroughStream)
+					else
+						rawStream = stream
+					# `stream` sends out raw strings in contrast to `task.progressHook`
+					# where we're given objects. capture these strings as they come
+					# before we parse them.
+					rawStream
+						.pipe(captureStream)
+						.pipe(buildProgressAdapter(inlineLogs))
+						.pipe(task.logStream)
+	.then (tasks) ->
+		logger.logDebug 'Prepared tasks; building...'
+		builder.performBuilds(tasks, docker)
+		.map (builtImage) ->
+			if not builtImage.successful
+				builtImage.error.serviceName = builtImage.serviceName
+				throw builtImage.error
+
+			d = imageDescriptorsByServiceName[builtImage.serviceName]
+			task = _.find(tasks, serviceName: builtImage.serviceName)
+
+			image =
+				serviceName: d.serviceName
+				name: d.image.tag ? d.image
+				logs: truncateString(task.logBuffer.join('\n'), LOG_LENGTH_MAX)
+				props:
+					dockerfile: builtImage.dockerfile
+					projectType: builtImage.projectType
+
+			# Times here are timestamps, so test whether they're null
+			# before creating a date out of them, as `new Date(null)`
+			# creates a date representing UNIX time 0.
+			if (startTime = builtImage.startTime)
+				image.props.startTime = new Date(startTime)
+			if (endTime = builtImage.endTime)
+				image.props.endTime = new Date(endTime)
+			docker.getImage(image.name).inspect().get('Size').then (size) ->
+				image.props.size = size
+			.return(image)
+		.tap (images) ->
+			summary = _(images).map ({ serviceName, props }) ->
+				[ serviceName, "Image size: #{humanize.filesize(props.size)}" ]
+			.fromPairs()
+			.value()
+			renderer.end(summary)
+	.finally(renderer.end)
+
+createRelease = (apiEndpoint, auth, userId, appId, composition) ->
+	_ = require('lodash')
+	crypto = require('crypto')
+	releaseMod = require('resin-release')
+	versions = require('./versions')
+
+	client = releaseMod.createClient({ apiEndpoint, auth })
+
+	releaseMod.create
+		client: client
+		user: userId
+		application: appId
+		composition: composition
+		source: "cli/#{versions.cli.major}"
```

Comment:
This is useful data, but it doesn't match the current format (which uses `local` for all CLI builds), which means right now on MC we don't show local builds as local in the UI.

What's the reason we want the CLI major version in here? We could put in a patch there to fix this fairly easily, but on closer thought I am a little surprised by this.

Reply:
Not sure TBH, I felt like it'd be nice to have a way to identify the origin with more granularity but I can't justify it anymore. I've changed it to "local" and removed the "version.coffee" file as well.

Reply:
I really like the idea of being able to specify the version of component which created the images/releases. Maybe this should be another field in the DB? (Of course this can be added after the initial release, as long as we make it optional.)